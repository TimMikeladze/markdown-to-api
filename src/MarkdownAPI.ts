import { globbySync } from 'globby';
import matter from 'gray-matter';
import slugify from 'slugify';
import MiniSearch, { Options as MiniSearchOptions } from 'minisearch';
import { remark } from 'remark';
import strip from 'strip-markdown';
import { existsSync, readFileSync, writeFileSync } from 'fs';
import { parse } from 'yaml';
import deepmerge from 'deepmerge';

const toSlug = (string: string): string => slugify(string, { lower: true, strict: true });

export interface FileMetadata {
  description?: string;
  id: string,
  path: string,
  slug?: string;
  tags?: string[];
  title?: string;
}

export interface ParsedFile {
  content: string;
  metadata: FileMetadata;
  strippedContent: string;
}

export interface Options {
  config?: any;
  configPath?: string;
  directory: string;
  indexPath?: string;
  miniSearchOptions?: MiniSearchOptions<any>;
  useIndex?: boolean;
  writeIndex?: boolean;
}

export interface Config {
  fields?: Record<string, {
    required?: boolean;
  } | null>,
  tags?: Record<string, {
    description?: string;
  } | null>
}

const defaultConfig: Config = {
  tags: {},
  fields: {
    title: {
      required: false,
    },
    description: {
      required: false,
    },
    tags: {
      required: false,
    },
  },
};

export class MarkdownAPI {
  private static readonly stripper = remark()
    .use(strip);

  private filePaths!: string[];

  private miniSearch!: MiniSearch;

  private files!: Map<string, ParsedFile>;

  private readonly options: Options;

  private config!: Config;

  constructor(options: Options) {
    this.options = options;
    this.init();
  }

  public getOptions(): Options {
    return this.options;
  }

  public init() {
    this.config = this.loadConfig();
    this.filePaths = this.loadFilePaths();
    this.files = this.loadFiles();
    if (this.options.useIndex && existsSync(this.getIndexPath())) {
      this.miniSearch = this.loadIndexJSON();
    } else {
      this.miniSearch = this.buildIndex();
    }
    if (this.options.writeIndex) {
      this.writeIndexJSON();
    }
  }

  public parseFile(path: string): ParsedFile {
    const output = matter(readFileSync(path, 'utf8'));

    let {
      data: {
        // eslint-disable-next-line prefer-const
        id, title, slug, tags, description,
      },
    }: {
      data: Partial<FileMetadata>;
    } = output;

    if (!title || !title.trim().length) {
      title = path.split('/').pop()?.split('.').shift();
    }

    if (!slug || !slug.trim().length) {
      slug = toSlug(title as string);
    }

    if (!id) {
      const parts = path.split('/');
      id = `${parts.slice(1, parts.length - 1).join('-')}-${slug}`;
    }

    id = toSlug(id);

    if (!description) {
      description = '';
    }

    const defaultFieldKeys = Object.keys(defaultConfig.fields || {});

    const extraFieldKeys = Object.keys(this.config.fields || {})
      .map((x) => (this.config.fields?.[x] && !defaultFieldKeys.includes(x) ? x : null))
      .filter(Boolean) as string[];

    const extraFields = extraFieldKeys.reduce<Record<string, string>>((res, key) => ({
      ...res,
      [key]: (output.data as any)[key],
    }), {});

    const strippedContent = (MarkdownAPI.stripper.processSync(output.content)).toString();

    return {
      metadata: {
        id,
        title,
        slug,
        path,
        tags,
        description,
        ...extraFields,
      },
      content: output.content,
      strippedContent,
    };
  }

  public loadConfig() {
    let config = {};
    if (this.options.configPath) {
      config = parse(readFileSync(this.options.configPath, 'utf8'));
    } else if (this.options.config) {
      config = this.options.config;
    } else if (existsSync(`${this.options.directory}/config.yml`)) {
      config = parse(readFileSync(`${this.options.directory}/config.yml`, 'utf8'));
    } else if (existsSync(`${this.options.directory}/config.yaml`)) {
      config = parse(readFileSync(`${this.options.directory}/config.yaml`, 'utf8'));
    }

    return deepmerge(defaultConfig, config);
  }

  public getConfig(): Config {
    return this.config;
  }

  public getTags(): {
    description?: string;
    name: string;
  }[] {
    return Object.keys(this.config.tags || {}).map((x) => ({
      name: x,
      description: this.config.tags?.[x]?.description || '',
    }));
  }

  public loadFilePaths() {
    const paths = globbySync(this.options.directory, {
      expandDirectories: {
        files: ['*.md'],
      },
    });
    return paths;
  }

  public loadFiles(): Map<string, ParsedFile> {
    const parsedFiles = this.filePaths.map((x) => this.parseFile(x));

    const requiredFields = Object.keys(this.config.fields || {})
      .map((x) => (this.config.fields?.[x]?.required ? x : null))
      .filter(Boolean);

    let errors: string[] = [];

    const files = parsedFiles.reduce((map, file) => {
      errors = [
        ...errors,
        ...requiredFields.reduce<string[]>((res, requiredField) => {
          if (requiredField && !((file.metadata as any)[requiredField])) {
            return [...res, `Missing required field '${requiredField}' in ${file.metadata.path}`];
          }
          return res;
        }, []),
      ];

      map.set(file.metadata.id, file);

      return map;
    }, new Map<string, ParsedFile>());

    if (errors.length) {
      throw new Error(errors.join('\n'));
    }

    return files;
  }

  // eslint-disable-next-line class-methods-use-this
  public getIndexFields() {
    const fields = ['title', 'slug', 'path', 'tags', 'description'];
    return {
      fields: [...fields, 'strippedContent'],
      storeFields: fields,
    };
  }

  public buildIndex(): MiniSearch {
    const miniSearch = new MiniSearch({
      ...this.getIndexFields(),
      ...this.options.miniSearchOptions,
    });

    miniSearch.addAll(Array.from(this.files.values()).map((x) => ({ ...x.metadata, ...x })));

    return miniSearch;
  }

  public getFile(id: string): ParsedFile | undefined {
    return this.files.get(id);
  }

  public getFiles(): Map<string, ParsedFile> {
    return this.files;
  }

  public getIndex() {
    return this.miniSearch;
  }

  public indexToJSON(): any {
    return JSON.stringify(this.miniSearch.toJSON());
  }

  public writeIndexJSON(): void {
    writeFileSync(this.getIndexPath(), this.indexToJSON(), 'utf8');
  }

  public getIndexPath(): string {
    return this.options.indexPath || `${this.options.directory}/index.json`;
  }

  public loadIndexJSON() {
    if (this.getIndexPath()) {
      return MiniSearch.loadJSON(readFileSync(this.getIndexPath(), 'utf8'), {
        ...this.getIndexFields(),
        ...this.options.miniSearchOptions,
      });
    }
    throw new Error('No index path provided');
  }

  public getFilePaths(): string[] {
    return this.filePaths;
  }
}
