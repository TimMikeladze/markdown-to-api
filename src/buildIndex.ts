import { globby } from 'globby';
import { readFile } from 'fs/promises';
import matter from 'gray-matter';
import slugify from 'slugify';
import MiniSearch from 'minisearch';
import { remark } from 'remark';
import strip from 'strip-markdown';

export const loadFilePaths = async (path: string) => {
  const paths = await globby(path, {
    expandDirectories: {
      files: ['*.md'],
    },
  });
  return paths;
};

export interface FileData {
  id: string,
  path: string,
  slug?: string;
  tags?: string[];
  title?: string;
}

export interface ParsedFile {
  content: string;
  data: FileData;
  rawContent: string;
}

const stripper = remark()
  .use(strip);

export const parseFile = async (path: string): Promise<ParsedFile> => {
  const output = matter(await readFile(path, 'utf8'));

  let {
    data: {
      // eslint-disable-next-line prefer-const
      id, title, slug, tags,
    },
  }: {
    data: Partial<FileData>;
  } = output;

  if (!title || !title.trim().length) {
    title = path.split('/').pop()?.split('.').shift();
  }

  if (!slug || !slug.trim().length) {
    slug = slugify(title as string, {
      lower: true,
      strict: true,
    });
  }

  if (!id) {
    const parts = path.split('/');
    id = `${parts.slice(0, parts.length - 1).join('/')}/${slug}`;
  }

  const rawContent = (await stripper.process(output.content)).toString();

  return {
    data: {
      id,
      title,
      slug,
      path,
      tags,
    },
    content: output.content,
    rawContent,
  };
};

export const parseFiles = (paths: string[]): Promise<ParsedFile[]> => {
  const files = Promise.all(paths.map(parseFile));
  return files;
};

export const buildIndex = async (files: ParsedFile[]) => {
  const fields = ['title', 'slug', 'path', 'tags'];

  const miniSearch = new MiniSearch({
    fields: [...fields, 'content'],
    storeFields: fields,
  });

  await miniSearch.addAllAsync(files.map((x) => ({ ...x.data, content: x.content })));

  return miniSearch;
};
