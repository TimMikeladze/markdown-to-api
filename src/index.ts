import { globby } from 'globby';
import { readFile } from 'fs/promises';
import matter from 'gray-matter';
import slugify from 'slugify';
import MiniSearch from 'minisearch';
import { remark } from 'remark';
import strip from 'strip-markdown';
import { existsSync, readFileSync } from 'fs';
import { parse } from 'yaml';
import deepmerge from 'deepmerge';

export interface FileData {
  description?: string;
  id: string,
  path: string,
  slug?: string;
  tags?: string[];
  title?: string;
}

export interface ParsedFile {
  content: string;
  data: FileData;
  strippedContent: string;
}

export interface Options {
  config?: any;
  configPath?: string;
  directory: string;
}

export interface Config {
  fields?: Record<string, {
    required?: boolean;
  } | null>,
  tags?: Record<string, {
    description?: string;
  } | null>
}

const defaultConfig: Config = {
  tags: {},
  fields: {
    title: {
      required: false,
    },
    description: {
      required: false,
    },
    tags: {
      required: false,
    },
  },
};

export class MarkdownToGraphQL {
  private static readonly stripper = remark()
    .use(strip);

  private filePaths!: string[];

  private miniSearch!: MiniSearch;

  private files!: Map<string, ParsedFile>;

  private readonly options: Options;

  private readonly config: Config;

  constructor(options: Options) {
    this.options = options;
    this.config = this.loadConfig();
  }

  public async init() {
    this.filePaths = await this.loadFilePaths();
    this.files = await this.loadFiles();
    this.miniSearch = await this.buildIndex();
  }

  public loadConfig() {
    let config = {};
    if (this.options.configPath) {
      config = parse(readFileSync(this.options.configPath, 'utf8'));
    } else if (this.options.config) {
      config = this.options.config;
    } else if (existsSync(`${this.options.directory}/config.yml`)) {
      config = parse(readFileSync(`${this.options.directory}/config.yml`, 'utf8'));
    } else if (existsSync(`${this.options.directory}/config.yaml`)) {
      config = parse(readFileSync(`${this.options.directory}/config.yaml`, 'utf8'));
    }

    return deepmerge(defaultConfig, config);
  }

  public getConfig(): Config {
    return this.config;
  }

  public static async parseFile(path: string): Promise<ParsedFile> {
    const output = matter(await readFile(path, 'utf8'));

    let {
      data: {
        // eslint-disable-next-line prefer-const
        id, title, slug, tags, description,
      },
    }: {
      data: Partial<FileData>;
    } = output;

    if (!title || !title.trim().length) {
      title = path.split('/').pop()?.split('.').shift();
    }

    if (!slug || !slug.trim().length) {
      slug = slugify(title as string, {
        lower: true,
        strict: true,
      });
    }

    if (!id) {
      const parts = path.split('/');
      id = `${parts.slice(0, parts.length - 1).join('/')}/${slug}`;
    }

    const strippedContent = (await MarkdownToGraphQL.stripper.process(output.content)).toString();

    return {
      data: {
        id,
        title,
        slug,
        path,
        tags,
        description,
      },
      content: output.content,
      strippedContent,
    };
  }

  public async loadFilePaths() {
    const paths = await globby(this.options.directory, {
      expandDirectories: {
        files: ['*.md'],
      },
    });
    return paths;
  }

  public async loadFiles(): Promise<Map<string, ParsedFile>> {
    const parsedFiles = await Promise.all(this.filePaths.map(MarkdownToGraphQL.parseFile));

    const requiredFields = Object.keys(this.config.fields || {}).map((x) => (this.config.fields?.[x]?.required ? x : null)).filter(Boolean);

    let errors: string[] = [];

    const files = parsedFiles.reduce((map, file) => {
      errors = [
        ...errors,
        ...requiredFields.reduce<string[]>((res, requiredField) => {
          if (requiredField && !((file.data as any)[requiredField])) {
            return [...res, `Missing required field '${requiredField}' in ${file.data.path}`];
          }
          return res;
        }, []),
      ];

      map.set(file.data.id, file);

      return map;
    }, new Map<string, ParsedFile>());

    if (errors.length) {
      throw new Error(errors.join('\n'));
    }

    return files;
  }

  public async buildIndex(): Promise<MiniSearch> {
    const fields = ['title', 'slug', 'path', 'tags', 'description'];

    const miniSearch = new MiniSearch({
      fields: [...fields, 'strippedContent'],
      storeFields: fields,
    });

    await miniSearch.addAllAsync(Array.from(this.files.values()).map((x) => ({ ...x.data, content: x.content })));

    return miniSearch;
  }

  public getFile(id: string): ParsedFile | undefined {
    return this.files.get(id);
  }

  public getFiles(): Map<string, ParsedFile> {
    return this.files;
  }

  public getIndex() {
    return this.miniSearch;
  }

  public getFilePaths(): string[] {
    return this.filePaths;
  }
}
