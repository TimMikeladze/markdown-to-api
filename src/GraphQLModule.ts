import { createModule, gql } from 'graphql-modules';
import { SearchOptions } from 'minisearch';
import { MarkdownAPI, Options } from './MarkdownAPI';

export const typeDefs = gql`
  type Query {
    countMarkdownFiles: Int
    markdownFile(id: ID!): MarkdownFile!
    markdownFiles(direction: MarkdownFilesSortDirection limit: Int offset: Int): [MarkdownFile!]!
    searchMarkdownFiles(
      text: String!
      options: MarkdownFileSearchOptions
      limit: Int
      offset: Int):
    MarkdownFileSearchResults!
    autoSuggestMarkdownFileSearch(text: String! options: MarkdownFileSearchOptions): MarkdownFileAutoSuggestResults!
    markdownFileTags: [MarkdownFileTag!]!
  }

  type MarkdownFile {
    content: String!
    strippedContent: String!
    id: ID!
    path: String!
    slug: String!
    tags: [String!]
    title: String!
    description: String
    createdAt: String!
  }

  type MarkdownFileSearchResults {
    count: Int!
    results: [MarkdownFileSearchResult!]!
  }

  type MarkdownFileSearchResult {
    id: ID!
    description: String
    score: Float!
    terms: [String!]
    title: String!
    slug: String!
    path: String!
    tags: [String!]
    createdAt: String!
    markdownFile: MarkdownFile!
  }

  type MarkdownFileTag {
    name: String!
    description: String
  }

  input MarkdownFileSearchOptions {
    fields: [String!]
    weights: MarkdownFileSearchOptionsWeights
    prefix: Boolean
    fuzzy: Boolean
    maxFuzzy: Int
    combineWith: MarkdownFileSearchOptionsCombineWith
  }

  type MarkdownFileAutoSuggestResults {
    count: Int!
    results: [MarkdownFileAutoSuggestResult!]!
  }

  type MarkdownFileAutoSuggestResult {
    suggestion: String!
    terms: [String!]!
    score: Float!
  }

  enum MarkdownFileSearchOptionsCombineWith {
    AND
    OR
  }

  input MarkdownFileSearchOptionsWeights {
    fuzzy: Float!
    exact: Float!
  }

  enum MarkdownFilesSortDirection {
    asc
    desc
  }

  enum MarkdownFilesSortBy {
    title
    createdAt
  }
`;

export const getResolvers = (mdapi: MarkdownAPI) => ({
  Query: {
    countMarkdownFiles() {
      return mdapi.countFiles();
    },
    markdownFileTags() {
      return mdapi.getTags();
    },
    markdownFile(_: any, { id }: { id: string }) {
      const file = mdapi.getFile(id);
      if (!file) {
        throw new Error(`File not found: ${id}`);
      }
      return file;
    },
    autoSuggestMarkdownFileSearch(_: any, { text, options }: { options?: SearchOptions, text: string; }) {
      const results = mdapi.getIndex().autoSuggest(text, {
        ...mdapi.getOptions(),
        ...options,
      });
      return {
        count: results.length,
        results,
      };
    },
    searchMarkdownFiles(_: any, {
      text, options, limit, offset,
    }: { limit?: number, offset?: number; options?: SearchOptions, text: string }) {
      const results = mdapi.getIndex().search(text, {
        ...mdapi.getOptions(),
        ...options,
      }).slice(offset || 0, limit);
      return {
        count: results.length,
        results,
      };
    },
    markdownFiles(
      _: any,
      {
        sortBy, direction, limit, offset,
      }:
        { direction: 'asc' | 'desc', limit?: number, offset?: number, sortBy: 'createdAt' | 'title' },
    ) {
      return mdapi.getFiles(sortBy, direction, limit, offset);
    },
  },
  MarkdownFileSearchResult: {
    markdownFile({ id }: { id: string }) {
      if (!id) {
        throw new Error('selection set is missing id');
      }
      const file = mdapi.getFile(id);
      return file;
    },
  },
});

export const createMarkdownAPIModule = (options: Options) => {
  const mdapi = new MarkdownAPI(options);

  return createModule({
    id: 'MarkdownAPIModule',
    typeDefs,
    resolvers: getResolvers(mdapi),
  });
};
