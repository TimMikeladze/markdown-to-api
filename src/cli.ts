import { Command } from 'commander';
import { MarkdownAPI } from './MarkdownAPI';

const program = new Command();

program
  .name('markdown-to-api')
  .description('CLI to interact with markdown-to-api');

program.requiredOption(
  '-d --directory <folder>',
  'Path to the folder containing the markdown files',
);

program.option(
  '-c --configPath <configPath>',
  'Path to the config file',
);

program.option(
  '-i --indexPath <indexPath>',
  'Path to the index file',
);

try {
  const res = program.parse(process.argv);

  const mdapi = new MarkdownAPI(res.opts());

  mdapi.writeIndexJSON();

  // eslint-disable-next-line no-console
  console.log(`Search index written to: ${mdapi.getIndexPath()}`);
} catch (err: any) {
  // eslint-disable-next-line no-console
  console.error(err.message);
}
