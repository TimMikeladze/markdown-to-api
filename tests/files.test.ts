import { MarkdownAPI } from '../src/MarkdownAPI';

describe('files', () => {
  let mdapi: MarkdownAPI;
  beforeAll(() => {
    mdapi = new MarkdownAPI({
      directory: 'tests/basic',
    });
  });

  it('return file map', () => {
    const res = mdapi.getFileMap();
    expect(res).toMatchSnapshot();
  });

  it('get all files title ascending', () => {
    const res = mdapi.getFiles('title', 'asc');
    expect(res).toMatchSnapshot();
  });

  it('get all files title descending', () => {
    const res = mdapi.getFiles('title', 'desc');
    expect(res).toMatchSnapshot();
  });

  it('get all files createdAt ascending', () => {
    const res = mdapi.getFiles('createdAt', 'asc');
    expect(res).toMatchSnapshot();
  });

  it('get all files createdAt descending', () => {
    const res = mdapi.getFiles('createdAt', 'desc');
    expect(res).toMatchSnapshot();
  });

  it('get all files with limit', () => {
    const res = mdapi.getFiles('createdAt', 'desc', 2);
    expect(res).toHaveLength(2);
  });

  it('get a file', () => {
    const res = mdapi.getFile('hello-world');
    expect(res).toBeDefined();
    expect(res).toMatchSnapshot();
  });

  it('does not return a file', () => {
    const res = mdapi.getFile('basic-dne');
    expect(res).toBeUndefined();
  });

  it('counts files', () => {
    expect(mdapi.countFiles()).toMatchSnapshot();
  });
});
