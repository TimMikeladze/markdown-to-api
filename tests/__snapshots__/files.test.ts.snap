// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`files counts files 1`] = `7`;

exports[`files get a file 1`] = `
Object {
  "content": "
# Hello world!
",
  "createdAt": "2022-08-13T00:03:28.460Z",
  "description": "",
  "id": "hello-world",
  "path": "tests/basic/hello.md",
  "slug": "hello-world",
  "strippedContent": "Hello world!
",
  "tags": Array [],
  "title": "Hello",
}
`;

exports[`files get all files createdAt ascending 1`] = `
Array [
  Object {
    "content": "
# Hello world!
",
    "createdAt": "2022-08-13T00:03:28.460Z",
    "description": "",
    "id": "hello-world",
    "path": "tests/basic/hello.md",
    "slug": "hello-world",
    "strippedContent": "Hello world!
",
    "tags": Array [],
    "title": "Hello",
  },
  Object {
    "content": "
# No title
",
    "createdAt": "2022-08-13T00:04:52.168Z",
    "description": "",
    "id": "no-title",
    "path": "tests/basic/no-title.md",
    "slug": "no-title",
    "strippedContent": "No title
",
    "tags": Array [],
    "title": "no-title",
  },
  Object {
    "content": "
# Nested File
",
    "createdAt": "2022-08-13T00:23:57.483Z",
    "description": "",
    "id": "nested-nested-file",
    "path": "tests/basic/nested/nested-file.md",
    "slug": "nested-file",
    "strippedContent": "Nested File
",
    "tags": Array [],
    "title": "Nested File",
  },
  Object {
    "content": "
# Tags!
",
    "createdAt": "2022-08-13T08:36:12.283Z",
    "description": "",
    "id": "with-tags",
    "path": "tests/basic/with-tags.md",
    "slug": "with-tags",
    "strippedContent": "Tags!
",
    "tags": Array [
      "tag1",
      "tag2",
    ],
    "title": "With tags",
  },
  Object {
    "content": "
# With description
",
    "createdAt": "2022-08-13T08:58:06.385Z",
    "description": "This is a description.",
    "id": "with-description",
    "path": "tests/basic/with-description.md",
    "slug": "with-description",
    "strippedContent": "With description
",
    "tags": Array [],
    "title": "With description",
  },
  Object {
    "content": "
# No slug or title
",
    "createdAt": "2022-08-18T19:41:31.436Z",
    "description": "",
    "id": "no-slug-or-title",
    "path": "tests/basic/no-slug-or-title.md",
    "slug": "no-slug-or-title",
    "strippedContent": "No slug or title
",
    "tags": Array [],
    "title": "no-slug-or-title",
  },
  Object {
    "content": "
# No slug
",
    "createdAt": "2022-08-18T19:41:31.440Z",
    "description": "",
    "id": "no-slug",
    "path": "tests/basic/no-slug.md",
    "slug": "no-slug",
    "strippedContent": "No slug
",
    "tags": Array [],
    "title": "No Slug",
  },
]
`;

exports[`files get all files createdAt descending 1`] = `
Array [
  Object {
    "content": "
# No slug
",
    "createdAt": "2022-08-18T19:41:31.440Z",
    "description": "",
    "id": "no-slug",
    "path": "tests/basic/no-slug.md",
    "slug": "no-slug",
    "strippedContent": "No slug
",
    "tags": Array [],
    "title": "No Slug",
  },
  Object {
    "content": "
# No slug or title
",
    "createdAt": "2022-08-18T19:41:31.436Z",
    "description": "",
    "id": "no-slug-or-title",
    "path": "tests/basic/no-slug-or-title.md",
    "slug": "no-slug-or-title",
    "strippedContent": "No slug or title
",
    "tags": Array [],
    "title": "no-slug-or-title",
  },
  Object {
    "content": "
# With description
",
    "createdAt": "2022-08-13T08:58:06.385Z",
    "description": "This is a description.",
    "id": "with-description",
    "path": "tests/basic/with-description.md",
    "slug": "with-description",
    "strippedContent": "With description
",
    "tags": Array [],
    "title": "With description",
  },
  Object {
    "content": "
# Tags!
",
    "createdAt": "2022-08-13T08:36:12.283Z",
    "description": "",
    "id": "with-tags",
    "path": "tests/basic/with-tags.md",
    "slug": "with-tags",
    "strippedContent": "Tags!
",
    "tags": Array [
      "tag1",
      "tag2",
    ],
    "title": "With tags",
  },
  Object {
    "content": "
# Nested File
",
    "createdAt": "2022-08-13T00:23:57.483Z",
    "description": "",
    "id": "nested-nested-file",
    "path": "tests/basic/nested/nested-file.md",
    "slug": "nested-file",
    "strippedContent": "Nested File
",
    "tags": Array [],
    "title": "Nested File",
  },
  Object {
    "content": "
# No title
",
    "createdAt": "2022-08-13T00:04:52.168Z",
    "description": "",
    "id": "no-title",
    "path": "tests/basic/no-title.md",
    "slug": "no-title",
    "strippedContent": "No title
",
    "tags": Array [],
    "title": "no-title",
  },
  Object {
    "content": "
# Hello world!
",
    "createdAt": "2022-08-13T00:03:28.460Z",
    "description": "",
    "id": "hello-world",
    "path": "tests/basic/hello.md",
    "slug": "hello-world",
    "strippedContent": "Hello world!
",
    "tags": Array [],
    "title": "Hello",
  },
]
`;

exports[`files get all files title ascending 1`] = `
Array [
  Object {
    "content": "
# Hello world!
",
    "createdAt": "2022-08-13T00:03:28.460Z",
    "description": "",
    "id": "hello-world",
    "path": "tests/basic/hello.md",
    "slug": "hello-world",
    "strippedContent": "Hello world!
",
    "tags": Array [],
    "title": "Hello",
  },
  Object {
    "content": "
# Nested File
",
    "createdAt": "2022-08-13T00:23:57.483Z",
    "description": "",
    "id": "nested-nested-file",
    "path": "tests/basic/nested/nested-file.md",
    "slug": "nested-file",
    "strippedContent": "Nested File
",
    "tags": Array [],
    "title": "Nested File",
  },
  Object {
    "content": "
# No slug
",
    "createdAt": "2022-08-18T19:41:31.440Z",
    "description": "",
    "id": "no-slug",
    "path": "tests/basic/no-slug.md",
    "slug": "no-slug",
    "strippedContent": "No slug
",
    "tags": Array [],
    "title": "No Slug",
  },
  Object {
    "content": "
# No slug or title
",
    "createdAt": "2022-08-18T19:41:31.436Z",
    "description": "",
    "id": "no-slug-or-title",
    "path": "tests/basic/no-slug-or-title.md",
    "slug": "no-slug-or-title",
    "strippedContent": "No slug or title
",
    "tags": Array [],
    "title": "no-slug-or-title",
  },
  Object {
    "content": "
# No title
",
    "createdAt": "2022-08-13T00:04:52.168Z",
    "description": "",
    "id": "no-title",
    "path": "tests/basic/no-title.md",
    "slug": "no-title",
    "strippedContent": "No title
",
    "tags": Array [],
    "title": "no-title",
  },
  Object {
    "content": "
# With description
",
    "createdAt": "2022-08-13T08:58:06.385Z",
    "description": "This is a description.",
    "id": "with-description",
    "path": "tests/basic/with-description.md",
    "slug": "with-description",
    "strippedContent": "With description
",
    "tags": Array [],
    "title": "With description",
  },
  Object {
    "content": "
# Tags!
",
    "createdAt": "2022-08-13T08:36:12.283Z",
    "description": "",
    "id": "with-tags",
    "path": "tests/basic/with-tags.md",
    "slug": "with-tags",
    "strippedContent": "Tags!
",
    "tags": Array [
      "tag1",
      "tag2",
    ],
    "title": "With tags",
  },
]
`;

exports[`files get all files title descending 1`] = `
Array [
  Object {
    "content": "
# Tags!
",
    "createdAt": "2022-08-13T08:36:12.283Z",
    "description": "",
    "id": "with-tags",
    "path": "tests/basic/with-tags.md",
    "slug": "with-tags",
    "strippedContent": "Tags!
",
    "tags": Array [
      "tag1",
      "tag2",
    ],
    "title": "With tags",
  },
  Object {
    "content": "
# With description
",
    "createdAt": "2022-08-13T08:58:06.385Z",
    "description": "This is a description.",
    "id": "with-description",
    "path": "tests/basic/with-description.md",
    "slug": "with-description",
    "strippedContent": "With description
",
    "tags": Array [],
    "title": "With description",
  },
  Object {
    "content": "
# No title
",
    "createdAt": "2022-08-13T00:04:52.168Z",
    "description": "",
    "id": "no-title",
    "path": "tests/basic/no-title.md",
    "slug": "no-title",
    "strippedContent": "No title
",
    "tags": Array [],
    "title": "no-title",
  },
  Object {
    "content": "
# No slug or title
",
    "createdAt": "2022-08-18T19:41:31.436Z",
    "description": "",
    "id": "no-slug-or-title",
    "path": "tests/basic/no-slug-or-title.md",
    "slug": "no-slug-or-title",
    "strippedContent": "No slug or title
",
    "tags": Array [],
    "title": "no-slug-or-title",
  },
  Object {
    "content": "
# No slug
",
    "createdAt": "2022-08-18T19:41:31.440Z",
    "description": "",
    "id": "no-slug",
    "path": "tests/basic/no-slug.md",
    "slug": "no-slug",
    "strippedContent": "No slug
",
    "tags": Array [],
    "title": "No Slug",
  },
  Object {
    "content": "
# Nested File
",
    "createdAt": "2022-08-13T00:23:57.483Z",
    "description": "",
    "id": "nested-nested-file",
    "path": "tests/basic/nested/nested-file.md",
    "slug": "nested-file",
    "strippedContent": "Nested File
",
    "tags": Array [],
    "title": "Nested File",
  },
  Object {
    "content": "
# Hello world!
",
    "createdAt": "2022-08-13T00:03:28.460Z",
    "description": "",
    "id": "hello-world",
    "path": "tests/basic/hello.md",
    "slug": "hello-world",
    "strippedContent": "Hello world!
",
    "tags": Array [],
    "title": "Hello",
  },
]
`;

exports[`files return file map 1`] = `
Map {
  "hello-world" => Object {
    "content": "
# Hello world!
",
    "createdAt": "2022-08-13T00:03:28.460Z",
    "description": "",
    "id": "hello-world",
    "path": "tests/basic/hello.md",
    "slug": "hello-world",
    "strippedContent": "Hello world!
",
    "tags": Array [],
    "title": "Hello",
  },
  "no-slug-or-title" => Object {
    "content": "
# No slug or title
",
    "createdAt": "2022-08-18T19:41:31.436Z",
    "description": "",
    "id": "no-slug-or-title",
    "path": "tests/basic/no-slug-or-title.md",
    "slug": "no-slug-or-title",
    "strippedContent": "No slug or title
",
    "tags": Array [],
    "title": "no-slug-or-title",
  },
  "no-slug" => Object {
    "content": "
# No slug
",
    "createdAt": "2022-08-18T19:41:31.440Z",
    "description": "",
    "id": "no-slug",
    "path": "tests/basic/no-slug.md",
    "slug": "no-slug",
    "strippedContent": "No slug
",
    "tags": Array [],
    "title": "No Slug",
  },
  "no-title" => Object {
    "content": "
# No title
",
    "createdAt": "2022-08-13T00:04:52.168Z",
    "description": "",
    "id": "no-title",
    "path": "tests/basic/no-title.md",
    "slug": "no-title",
    "strippedContent": "No title
",
    "tags": Array [],
    "title": "no-title",
  },
  "with-description" => Object {
    "content": "
# With description
",
    "createdAt": "2022-08-13T08:58:06.385Z",
    "description": "This is a description.",
    "id": "with-description",
    "path": "tests/basic/with-description.md",
    "slug": "with-description",
    "strippedContent": "With description
",
    "tags": Array [],
    "title": "With description",
  },
  "with-tags" => Object {
    "content": "
# Tags!
",
    "createdAt": "2022-08-13T08:36:12.283Z",
    "description": "",
    "id": "with-tags",
    "path": "tests/basic/with-tags.md",
    "slug": "with-tags",
    "strippedContent": "Tags!
",
    "tags": Array [
      "tag1",
      "tag2",
    ],
    "title": "With tags",
  },
  "nested-nested-file" => Object {
    "content": "
# Nested File
",
    "createdAt": "2022-08-13T00:23:57.483Z",
    "description": "",
    "id": "nested-nested-file",
    "path": "tests/basic/nested/nested-file.md",
    "slug": "nested-file",
    "strippedContent": "Nested File
",
    "tags": Array [],
    "title": "Nested File",
  },
}
`;
